### PR TITLE
District lookup

### PR DESCRIPTION
**Pull request by turino**
Basic district lookup with tests.

I put my controller spec in spec/controllers. Looks like other controller specs are in spec/requests. These should probably be in same place?

I haven't written any model tests yet, but also haven't written any methods in models. I want to refactor districts_controller, maybe moving code to models. If so, I'll write model tests.

I don't know what to do about line 8 in mobile_commons.rb:
```
SINGLE_DISTRICT_STATES = ["AK", "DE", "MT", "ND", "SD", "VI", "VT", "WY", "AS", "DC", "GU", "MP", "PR"]

### CODE REVIEW

**Comment by turino** [comment]
oops -- accidentally submitted before I meant to.

I was saying that I want to come up with better way to handle single-district states
```
SINGLE_DISTRICT_STATES = ["AK", "DE", "MT", "ND", "SD", "VI", "VT", "WY", "AS", "DC", "GU", "MP", "PR"]
```
seems silly to hard-code it, but didn't really want to have to pass it in as a parameter every time. (This list of states probably won't change any time soon)

File: app/controllers/v1/districts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        else
+          targeted = false
+        end
+      end
+      { targeted: targeted, district: district, city: city, state: state }
+    end
+
+    def mcommons_district(coords)
+      results = Integration::MobileCommons.district_from_coords(coords)
+      state = State.find_by(abbrev: results[:state])
```
Comment (Rio517) [inline app/controllers/v1/districts_controller.rb]: Not sure you need to go through state here - why not use the state provided by Mcommons?

File: app/controllers/v1/districts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      render json: { address:    results[:address_name],
+                     confidence: results[:confidence],
+                     coords: coords
+                   }.merge(district_info)
+    
+    elsif zip = params[:zip].presence
+      render json: district_info_for_zip(zip),
+              status: 200
+    else
+      render json: { error: 'zip code is required' }
```
Comment (Rio517) [inline app/controllers/v1/districts_controller.rb]: I think your test for this error is address, not zip.

Comment (Rio517) [inline app/controllers/v1/districts_controller.rb]: Oh. I see.  I didn't read far enough.  all good.

File: app/controllers/v1/districts_controller.rb (hunk 1 above)
Comment (turino) [inline app/controllers/v1/districts_controller.rb]: I think you're right. That was left over from previous implementation. I'll fix it

Comment (turino) [inline app/controllers/v1/districts_controller.rb]: Actually it is needed. 

You're talking about line 43, right? I need the state_id in order to find the district

**Comment by Rio517** [comment]
This is generally fine, but I'd make a few notes:

* don't worry about hard coding at-large districts - as their called.  That was a mistake made before I was at mobile commons - using 1 instead of 0 as the district identifier.  Unfortunately, there really aren't many work arounds.
* I do think you have a lot of logic in your controller - i'd extract it into either the models, presenters or your integration class - if you're just parsing data from an integration.
* you're calling `render json:` multiple times as you build possible hash outputs. I'd build the hash then just make that call once at the end. 
* For managing logic, you could try using a [presentre](https://quickleft.com/blog/presenters-as-a-solution-to-asjson-woes-in-rails-apis/) or [jbuilder](https://github.com/rails/jbuilder).
* You're right - the specs should be in the controller folder - I was original going to go with header params for the API versioning, but decided to go with path. In the header case, controller specs won't work w/ the gem I'm using.  I'll move them on master. 

I'm going to go ahead and merge. feel free to make more edits on this branch and we can merge them in later.

Thanks, 

-Mario

**Pull request merged**